core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp

```diff
@@ -130,18 +149,41 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
     auto result_ptr = static_cast<pointer_type>(
         sycl::malloc(sizeof(*m_result_ptr), q, sycl::usm::alloc::shared));
 
-    value_type host_result;
-    ValueInit::init(functor, &host_result);
+    value_type identity{};
+    if constexpr (!std::is_same<ReducerType, InvalidType>::value)
+      m_reducer.init(identity);
+
+    value_type host_result = identity;
+    if constexpr (ReduceFunctorHasInit<Functor>::value)
+      ValueInit::init(functor, &host_result);
     q.memcpy(result_ptr, &host_result, sizeof(host_result));
```

Comment:
I think we should just use assignment here `*result_ptr = host_result;`, mainly because the rest of the code also takes advantage of it being in USM shared memory (`*m_result_ptr = *result_ptr;`, for instance).

Reply:
Improved in [`e297544` (#3544)](https://github.com/kokkos/kokkos/pull/3544/commits/e297544bc70f3966eb31047a9eec75ddb5417a44)